service/storage_service.hh

```diff
@@ -750,8 +750,15 @@ private:
         // The node uses raft-based topology operations
         raft
     };
+    // The _topology_change_kind_enabled variable is first initialized in `join_cluster`.
+    // After the node successfully joins, the control over the variable is yielded
+    // to `topology_state_load`, so that it can control it during the upgrade from gossiper
+    // based topology to raft-based topology.
+    bool _manage_topology_change_kind_from_group0 = false;
```

Comment:
This is a hack to work around the following issue: when a node creates a new cluster, it starts group 0 with an empty state. If `topology_state_load` were to always update this field then it would temporarily revert `_topology_change_kind_enabled` to `legacy`.

I'll try to fix this by initializing the newly created group 0 during `setup_group0` (effectively move `raft_initialize_discovery_leader` to happen during it through the handshaker).

I'll also try to reorganize the beginning of `join_token_ring` so that the mode we use is implied by the control flow and not `_topology_change_kind_enabled`, at least not until gossiping starts (because it reads the flag).

Reply:
Gleb's arguments in the other thread convinced me that it's not worth implementing the blocking semantics for `upgrade` (needs an additional step), so I'm resolving.

Reply:
I added a FIXME comment to the code.